### PR TITLE
Add ability to set custom menu background in `SkinEditor`

### PR DESCRIPTION
**Pull request by acid-chicken**
This change proposes to allow osu!supporters to set custom menu backgrounds even for argon skins and triangle skins.

https://user-images.githubusercontent.com/20679825/198125457-36fe3a59-1a62-4958-9d90-5c605f9e63bf.mp4

### CODE REVIEW

**Comment by bdach** [comment]
It is generally recommended to discuss feature proposals as large as this beforehand. Not sure why this hasn't been done in this case.

File: osu.Game/Skinning/Editor/SkinEditor.cs [hunk 1]
```diff
@@ -93,93 +102,106 @@ private void load()
...
-                                                },
+                                                    new MenuItem("File")
+                                                    {
+                                                        Items = new[]
+                                                        {
+                                                            new EditorMenuItem("Save", MenuItemType.Standard, Save),
+                                                            new EditorMenuItem("Revert to default", MenuItemType.Destructive, revert),
+                                                            new EditorMenuItemSpacer(),
+                                                            new EditorMenuItem("Set meun background", MenuItemType.Standard, () => menuBackgroundChooser.ShowPopover()),
+                                                            new EditorMenuItem("Remove meun background", MenuItemType.Standard, removeMenuBackground),
```
Comment (bdach) [inline osu.Game/Skinning/Editor/SkinEditor.cs:150]: typos

```suggestion
                                                            new EditorMenuItem("Set menu background", MenuItemType.Standard, () => menuBackgroundChooser.ShowPopover()),
                                                            new EditorMenuItem("Remove menu background", MenuItemType.Standard, removeMenuBackground),
```

File: osu.Game/Skinning/Editor/SkinEditor.cs [hunk 2]
```diff
@@ -330,6 +352,68 @@ private void revert()
...
+                stream.CopyTo(menuBackgroundStream);
+            menuBackgroundStream.Seek(0, SeekOrigin.Begin);
+
+            Texture previewTexture = Texture.FromStream((skins as IO.IStorageResourceProvider).Renderer, menuBackgroundStream);
+            menuBackgroundStream.Seek(0, SeekOrigin.Begin);
+
+            currentSkin.Value.SkinInfo.PerformWrite(skinInfo =>
+            {
+                using (menuBackgroundStream)
+                    skins.AddFile(skinInfo, menuBackgroundStream, $"menu-background{fileInfo.Extension}");
```
Comment (bdach) [inline osu.Game/Skinning/Editor/SkinEditor.cs:379]: Very strange that the `menuBackgroundStream` is disposed twice (once in the outer scope, once in the inner scope here)

File: osu.Game/Skinning/Editor/SkinEditor.cs [hunk 3]
```diff
@@ -330,6 +352,68 @@ private void revert()
...
+            });
+            realm.Run(r => r.Refresh());
```
Comment (bdach) [inline osu.Game/Skinning/Editor/SkinEditor.cs:381]: What is the purpose of this refresh? I believe that non-async realm writes already implicitly trigger refreshes?

File: osu.Game/Skinning/Editor/SkinEditor.cs [hunk 4]
```diff
@@ -330,6 +352,68 @@ private void revert()
...
+            if (previewBackground != null)
+                Remove(previewBackground, true);
+
+            onScreenDisplay?.Display(new SkinEditorToast(ToastStrings.MenuBackgroundChanged, fileInfo.Name));
+            game?.PerformFromScreen(screen =>
+            {
+                (screen as MainMenu)?.ApplyToBackground(b =>
+                {
+                    if (b is BackgroundScreenDefault target)
+                        AddInternal(previewBackground = new MenuBackgroundPreview(target, previewTexture));
```
Comment (bdach) [inline osu.Game/Skinning/Editor/SkinEditor.cs:392]: I don't understand the point of this "background preview" thing. The skin is being changed instantaneously, so why bother with adding so much complexity? Not only you have to add and remove this "preview", you also have to "lock" and "unlock" the background. This seems too complicated for what it really needs to be.

In my ideal implementation of this:

* `MenuBackgroundPreview` does not exist
* `BackgroundScreenDefault.{Lock,Unlock}()` do not exist
* `BackgroundScreenDefault.GetBackgroundFallbackTextureName()` is private again

Unless a very compelling reason is provided for structuring things this way I can't find myself agreeing with any of this.

File: osu.Game/Skinning/Editor/SkinEditor.cs [hunk 5]
```diff
@@ -343,6 +427,7 @@ public void Save()
                 currentSkin.Value.UpdateDrawableTarget(t);
 
             skins.Save(skins.CurrentSkin.Value);
+            skins.Reload();
```
Comment (bdach) [inline osu.Game/Skinning/Editor/SkinEditor.cs:430]: Why is this added?

File: osu.Game/Skinning/Editor/SkinEditor.cs [hunk 6]
```diff
@@ -379,8 +467,16 @@ public void DeleteItems(ISkinnableDrawable[] items)
 
         public Task Import(params string[] paths)
         {
+            if (menuBackgroundChooser.PopoverVisible)
+                return Task.CompletedTask;
```
Comment (bdach) [inline osu.Game/Skinning/Editor/SkinEditor.cs:471]: This early-return needs an explanatory comment as to why it exists at the least. But I find it really strange that this PR adds *two* import handlers, one of which just plays some sounds and closes the import popover, and the other does the actual import logic outside of the component responsible for actually selecting the background file.

File: osu.Game/Skinning/Editor/SkinEditorMenuBackgroundChooser.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        public Task Import(params string[] paths)
+        {
+            Schedule(() =>
+            {
+                HoverClickSounds sounds = new HoverClickSounds();
+                AddInternal(sounds);
+                sounds.TriggerClick();
+                sounds.Expire();
```
Comment (bdach) [inline osu.Game/Skinning/Editor/SkinEditorMenuBackgroundChooser.cs:62]: On an audio design level I don't think drag-and-drop operations should be emitting click sounds.

**Review by bdach — COMMENTED** [review]
Here's a rough review. There is a lot of complexity in this and I don't know why it's there as it's not explained anywhere.

**Comment by bdach** [comment]
Also test failures requiring attention.

**Comment by peppy** [comment]
Thanks for your contribution.

Just reading a few lines into the code here, I don't think this is the correct approach.

This should be implemented using the standard `ISkinnableDrawable` flow, with the texture being chosen as a setting, similar to other customisable skin elements with textures. We can't have inidividual elements like this adding new menu options and UI elements. Everything should follow a single flow.

I already have experimented with adding customisation of the background blur at song select, which would be similar to how I would expect this to be implemented.

To hook everything up well, the skin layout editor needs better support for different scenes, and the texture dropdown needs a way to add new textures that isn't drag-drop. All these pieces should be implemented in small steps, to make end goals like that proposed in this PR possible.

**Comment by acid-chicken** [comment]
Thank you for the review.

First, I apologize that I should have first calmly analyzed the fact that the patch has become more huge than I initially expected.

> In my ideal implementation of this:
>
> - `MenuBackgroundPreview` does not exist
> - `BackgroundScreenDefault.{Lock,Unlock}()` do not exist
> - `BackgroundScreenDefault.GetBackgroundFallbackTextureName()` is private again
> - All changes are made on the live skin, so after the background change is done, all you have to do is call `.Next()`

In particular, I completely agree with this one.
The reason for this situation is that textures in `IBindable<Skin>` are still not updated even after running `SkinEditor.Save(Skin)`.
I explore related commits and find this problem had manifested itself in the past (e.g. https://github.com/ppy/osu/compare/master...peppy:osu:skin-editor-replace-textures#diff-60a9b9b58f1dac44e878747e45dce0532d82f6e8f42e6cde4ae9a786eb7a2803R116-R122), so, either way, I will try to solve this problem first, whether implementing this feature.

**Comment by peppy** [comment]
Yeah, approaching in small pieces is definitely recommended. And don't hesitate to ask questions or propose solutions (here or discord) before writing the code if you want more immediate feedback.
